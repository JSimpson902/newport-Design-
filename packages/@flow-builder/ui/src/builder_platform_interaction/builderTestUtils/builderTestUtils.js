export * from './screenEditorTestUtils.js';
export * from './commonTestUtils';
export * from './deepFindMatchers';
export * from './useAdvancedOptionsTestUtils';
export * from './goldObjectMatchers';
export * from './domTestUtils';
export * from './complexTypeTestUtils';
