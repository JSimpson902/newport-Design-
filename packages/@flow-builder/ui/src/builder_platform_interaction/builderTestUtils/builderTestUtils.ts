export * from './combobox';
export * from './commonTestUtils';
export * from './complexTypeTestUtils';
export * from './deepFindMatchers';
export * from './domTestUtils';
export * from './events';
export * from './goldObjectMatchers';
export * from './manuallyAssignVariablesCheckboxTestUtils';
export * from './screenEditorTestUtils';
export * from './selectors';

// @sa11y testing appears to take some extra time (beyond the given 5000ms default jest timeout)
export const ACCESSIBILITY_TEST_TIMEOUT_IN_MS = 15000;

/**
 * Mocks a label in jest
 *
 * @param jest - The jest instance
 * @param labelName - The label name
 * @returns the mock label
 */
export function mockLabel(jest, labelName: string) {
    jest.mock(
        `@salesforce/label/${labelName}`,
        () => {
            return { default: `${labelName} mock value` };
        },
        { virtual: true }
    );

    return `${labelName} mock value`;
}
