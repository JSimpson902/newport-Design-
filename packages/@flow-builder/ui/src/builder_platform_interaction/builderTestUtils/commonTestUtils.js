const DEFAULT_MAX_TICKS = 10;

/**
 * Returns a promise that is fulfilled when condition is truthy and rejected if we exceed given max number of event loop ticks
 *
 * @param {Function} condition should return truthy value when condition is met
 * @param {number} maxTicks max number of event loop ticks before the condition becomes truthy
 * @returns {Promise} if fulfilled, value is the condition returned value
 */
export const until = (condition, maxTicks = DEFAULT_MAX_TICKS) => {
    return Promise.resolve().then(() => {
        try {
            const result = condition();
            if (result) {
                return Promise.resolve(result);
            }
            if (maxTicks === 0) {
                return Promise.reject(Error('Condition not met'));
            }
        } catch (error) {
            if (maxTicks === 0) {
                return Promise.reject(error);
            }
        }
        return until(condition, maxTicks - 1);
    });
};

/**
 * Returns a promise that is fulfilled when expectations are met
 * @param {Function} expectations should throw an exception if an expectation is not met
 * @param {number} maxTicks max number of event loop ticks before the condition becomes truthy
 * @returns {Promise} fulfilled when expectations are met
 */
export const untilNoFailure = (expectations, maxTicks = DEFAULT_MAX_TICKS) => {
    return until(() => {
        expectations();
        return true;
    }, maxTicks);
};

/**
 * Returns a promise that is fulfilled after the given number of event loop ticks
 * @param {number} maxTicks number of event loop ticks before the promise is fulfilled
 * @returns {Promise} fulfilled after the given number of event loop ticks
 */
export const ticks = (maxTicks = DEFAULT_MAX_TICKS) => {
    return until(() => false, maxTicks).catch(() => {});
};