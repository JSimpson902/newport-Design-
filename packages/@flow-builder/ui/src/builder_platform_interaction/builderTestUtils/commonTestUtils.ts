import { createElement } from 'lwc';
import { setDocumentBodyChildren } from './domTestUtils';

const DEFAULT_MAX_TICKS = 10;

/**
 * Returns a promise that is fulfilled when condition is truthy and rejected if we exceed given max number of event loop ticks
 *
 * @param {Function} condition should return truthy value when condition is met
 * @param {number} maxTicks max number of event loop ticks before the condition becomes truthy
 * @returns {Promise} if fulfilled, value is the condition returned value
 */
export const until = (condition, maxTicks = DEFAULT_MAX_TICKS) =>
    Promise.resolve().then(() => {
        try {
            const result = condition();
            if (result) {
                return Promise.resolve(result);
            }
            if (maxTicks === 0) {
                return Promise.reject(Error('Condition not met'));
            }
        } catch (error) {
            if (maxTicks === 0) {
                return Promise.reject(error);
            }
        }
        return until(condition, maxTicks - 1);
    });

/**
 * Returns a promise that is fulfilled when expectations are met
 *
 * @param {Function} expectations should throw an exception if an expectation is not met
 * @param {number} maxTicks max number of event loop ticks before the condition becomes truthy
 * @returns {Promise} fulfilled when expectations are met
 */
export const untilNoFailure = (expectations, maxTicks = DEFAULT_MAX_TICKS) =>
    until(() => {
        expectations();
        return true;
    }, maxTicks);

/**
 * Returns a promise that is fulfilled after the given number of event loop ticks
 *
 * @param {number} maxTicks number of event loop ticks before the promise is fulfilled
 * @returns {Promise} fulfilled after the given number of event loop ticks
 */
export const ticks = (maxTicks = DEFAULT_MAX_TICKS) => until(() => false, maxTicks).catch(() => {});

/**
 * Utility function that waits 1 tick (can be used to wait for rendering after a component update)
 */
export async function tick() {
    await ticks(1);
}

/**
 * Updates a lightning component's props and waits for it to render
 *
 * @param cmp - The component
 * @param props - The props
 */
export async function updateComponent(cmp, props) {
    Object.assign(cmp, props);
    await tick();
}

/**
 * Create a new promise with status properties.
 *
 * @param promise current promise
 * @returns new promise with status properties
 */
export const makeQuerablePromise = (promise) => {
    // Don't modify any promise that has been already modified.
    if (promise.isResolved) {
        return promise;
    }

    let isPending = true;
    let isRejected = false;
    let isFulfilled = false;

    const result = promise.then(
        (value) => {
            isFulfilled = true;
            isPending = false;
            return value;
        },
        (e) => {
            isRejected = true;
            isPending = false;
            throw e;
        }
    );

    result.isFulfilled = () => isFulfilled;
    result.isPending = () => isPending;
    result.isRejected = () => isRejected;
    return result;
};

/**
 * Formats a string to camelcase
 *
 * @param s - the string to format
 * @returns the formatted string
 */
const camelCase = (s: string) => s.replace(/-[a-zA-Z]/gi, (match) => match.charAt(1).toUpperCase());

/**
 * Creates a component for use in jest tests
 *
 * @param tagName - The component name (with namespace eg: builder_platform_interaction-breadcrumbs)
 * @param options - The component properties
 * @param optionsOverride - The overridden component properties
 * @returns promise with the lightning element for the component as returned value
 */
export const createComponent = async (tagName, options = {}, optionsOverride = {}) => {
    options = { ...options, ...optionsOverride };

    const [namespace] = tagName.split('-');
    const moduleName = camelCase(tagName.replace(`${namespace}-`, ''));

    const el = createElement(tagName, {
        is: require(`${namespace}/${moduleName}`).default
    });

    Object.assign(el, options);

    setDocumentBodyChildren(el);
    await tick();
    return el;
};

/*
 * @param callback Jest mocked callback
 * @returns the mocked callback event detail object value
 */
export const getDetailPassedToEvent = (callback: { mock: { calls: { detail: {} }[][] } }) =>
    callback.mock.calls[0][0].detail;
