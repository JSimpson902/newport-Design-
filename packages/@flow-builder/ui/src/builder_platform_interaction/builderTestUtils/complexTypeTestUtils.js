import { FLOW_DATA_TYPE } from 'builder_platform_interaction/dataTypeLib';

export const expectComplexTypeFieldDescription = field => {
    // need a dataType and apiName. isCollection and label optional
    expect(field.dataType).toBeDefined();
    expect(field.apiName).toBeDefined();
    if (
        field.dataType === FLOW_DATA_TYPE.SOBJECT.value ||
        field.dataType === FLOW_DATA_TYPE.APEX
    ) {
        expect(field.subtype).toBeDefined();
    }
};

export const expectFieldsAreComplexTypeFieldDescriptions = fields => {
    for (const fieldName in fields) {
        if (Object.prototype.hasOwnProperty.call(fields, fieldName)) {
            const field = fields[fieldName];
            expectComplexTypeFieldDescription(field);
        }
    }
};
