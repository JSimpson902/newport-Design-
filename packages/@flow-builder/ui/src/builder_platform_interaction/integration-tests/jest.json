{
  "moduleNameMapper": {
    "^mock-store-data$": "<rootDir>/../../../jest-mock-data/mock_store_data/mock-store-data.js",
    "^mock-server-entity-data$": "<rootDir>/../../../jest-mock-data/mock_server_entity_data/mock-server-entity-data.js",
    "^builder_platform_interaction-store-lib$": "<rootDir>/../../../jest-modules/builder_platform_interaction/store-lib/store-lib-mock.js",    
    "^(builder_platform_interaction)-(.+)$": "<rootDir>/../../../modules/$1/$2/$2",
    "^(lightning)-(.+)$": "<rootDir>/../../../jest-modules/$1/$2/$2"
  },
  "setupTestFrameworkScriptFile": "<rootDir>/../../../jestSetup.js",
  "clearMocks": true
}