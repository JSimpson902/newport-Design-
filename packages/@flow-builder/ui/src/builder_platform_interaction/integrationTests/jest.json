{
  "moduleNameMapper": {
    "^mock/ruleService$": "<rootDir>/../../../jest-mock-data/mock_rule_service/mock-rule-service.js",
    "^mock/storeData$": "<rootDir>/../../../jest-mock-data/mock_store_data/mock-store-data.js",
    "^mock/serverEntityData$": "<rootDir>/../../../jest-mock-data/mock_server_entity_data/mock-server-entity-data.js",
    "^builder_platform_interaction/storeLib$": "<rootDir>/../../../jest-modules/builder_platform_interaction/storeLib/storeLib-mock.js",    
    "^(builder_platform_interaction)/(.+)$": "<rootDir>/../../../modules/$1/$2/$2",
    "^(lightning)/(.+)$": "<rootDir>/../../../jest-modules/$1/$2/$2"
  },
  "setupTestFrameworkScriptFile": "<rootDir>/../../../jestSetup.js",
  "clearMocks": true
}