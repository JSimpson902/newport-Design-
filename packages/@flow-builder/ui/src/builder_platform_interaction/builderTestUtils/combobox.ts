import Combobox from 'builder_platform_interaction/combobox';
import { clickEvent, LIGHTNING_COMPONENTS_SELECTORS, removeEvent, ticks } from './builderTestUtils';

/**
 * Return the pill component corresponding to the given combobox
 *
 * @param {Combobox} combobox - combobox component
 * @returns {HTMLElement | null} lightning pill component
 */
export const getComboboxPill = (combobox: any): any | null => {
    const { shadowRoot } = combobox;
    if (!shadowRoot) {
        throw new Error('combobox shadowRoot must not be null');
    }
    return shadowRoot.querySelector(LIGHTNING_COMPONENTS_SELECTORS.LIGHTNING_PILL);
};

/**
 * Click on pill to switch to merge field notation combobox mode
 *
 * @param {Combobox} combobox - current combobox
 * @returns {Promise<void>}
 */
// eslint-disable-next-line @lwc/lwc/no-async-await
export const clickPill = async (combobox: Combobox): Promise<void> => {
    const pill = getComboboxPill(combobox);
    if (pill) {
        pill.dispatchEvent(clickEvent());
        await ticks(1);
    }
};

/**
 * Remove pill to switch to merge field notation combobox mode and reset combobox menu data
 *
 * @param {Combobox} combobox - current combobox
 * @returns {Promise<void>}
 */
// eslint-disable-next-line @lwc/lwc/no-async-await
export const removePill = async (combobox: Combobox): Promise<void> => {
    const pill = getComboboxPill(combobox);
    if (pill) {
        pill.dispatchEvent(removeEvent());
        await ticks(1);
    }
};
