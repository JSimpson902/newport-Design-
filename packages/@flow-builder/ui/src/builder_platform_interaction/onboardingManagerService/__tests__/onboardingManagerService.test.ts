// @ts-nocheck
import { setup } from '@sa11y/jest';
import { createComponent } from 'builder_platform_interaction/builderTestUtils';

const createComponentUnderTest = async () => {
    return createComponent('builder_platform_interaction-onboarding-manager-service');
};

describe('onboardingManagerService', () => {
    beforeAll(() => {
        setup();
    });

    afterEach(() => {
        while (document.body.firstChild) {
            document.body.removeChild(document.body.firstChild);
        }
    });

    test('renders', async () => {
        const cmp = await createComponentUnderTest();
        expect(cmp).toMatchSnapshot();
    });

    it('is accessible', async () => {
        const cmp = await createComponentUnderTest();
        expect(cmp).toBeAccessible();
    });

    describe('implements required APIs', () => {
        it('getAppContainerId set', async () => {
            const service = await createComponentUnderTest();
            expect(service.getAppContainerId()).toBe('flowbuilder');
        });

        it('getPromptRegistrations set', async () => {
            const service = await createComponentUnderTest();
            expect(service.getPromptRegistrations()).toBeDefined();
        });
    });
});
