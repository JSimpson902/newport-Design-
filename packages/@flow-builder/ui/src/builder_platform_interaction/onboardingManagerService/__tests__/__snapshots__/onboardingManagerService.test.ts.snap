// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`onboardingManagerService renders 1`] = `
<builder_platform_interaction-onboarding-manager-service>
  #shadow-root(open)
</builder_platform_interaction-onboarding-manager-service>
`;
